lib/resty/healthcheck.lua

```diff
@@ -1690,9 +1690,11 @@ function _M.new(opts)
         end
 
         local cur_time = ngx_now()
+        local is_checked = false
         for _, checker_obj in pairs(hcs) do
 
           if (last_cleanup_check + CLEANUP_INTERVAL) < cur_time then
+            is_checked = true
```

Comment:
I don't get how this change fixes the issue. It seems the same code for me. Did you get to consistently reproduce the bug?

Reply:
Please be careful this LOC https://github.com/Kong/lua-resty-healthcheck/blob/master/lib/resty/healthcheck.lua#L1725

```
        for _, checker_obj in pairs(hcs) do

          if (last_cleanup_check + CLEANUP_INTERVAL) < cur_time then
            -- clear targets marked for delayed removal
            ****
            last_cleanup_check = cur_time
          end

          if checker_obj.checks.active.healthy.active and
            (checker_obj.checks.active.healthy.last_run +
              checker_obj.checks.active.healthy.interval <= cur_time)
          then
            checker_obj.checks.active.healthy.last_run = cur_time
            checker_callback(checker_obj, "healthy")
          end

          if checker_obj.checks.active.unhealthy.active and
            (checker_obj.checks.active.unhealthy.last_run +
              checker_obj.checks.active.unhealthy.interval <= cur_time)
          then
            checker_obj.checks.active.unhealthy.last_run = cur_time
            checker_callback(checker_obj, "unhealthy")
          end
        end
```

If we check the first healthcheck instance delayed removed target and replace `last_cleanup_check` with the current time. and then start checking the second healthcheck instance, it will skip because the first health check instance check logic has been executed `last_cleanup_check = cur_time`.

Reply:
Gotcha, it went over my head! Thanks for making it clear.

Reply:
@oowl I got this problem too and I fixed it as you say.But there are targets left that not been cleared by delay clear when acquiring lock timeout.

```
2023/12/19 19:30:19 [error] 23583#23583: *467115 [lua] healthcheck.lua:1385: log(): [healthcheck] (upstream#upstreams/1) cleanup check failed, err: failed acquiring lock for 'lua-resty-healthcheck:upstream#upstreams/1:target_list_lock', timeout, context: ngx.timer, client: xxx, server: 0.0.0.0:80
```

the log to print the error

![image](https://github.com/Kong/lua-resty-healthcheck/assets/16033853/2ec8f854-a1bf-4fec-8a38-497290efcd98)

Reply:
That's nice catch, I have resolved it. Please check again.